server/models/contests/participate.js

```diff
@@ -0,0 +1,77 @@
+const { pool } = require('../database')
+
+/**
+ * @param {*} param0
+ * @param {Object} param0.params
+ * @param {String} param0.username
+ * @return {Promise}
+ */
+
+function participate({ params, username }) {
+  return new Promise((resolve, reject) => {
+    const { contest_id: contestId } = params
+    pool.getConnection((error, connection) => {
+      if (error) {
+        return reject(error)
+      }
+      connection.beginTransaction((error) => {
+        if (error) {
+          connection.release()
+          return reject(error)
+        }
+        connection.query(
+          `INSERT INTO contests_participants (contest_id, participant) SELECT ?, ? 
+          WHERE EXISTS(SELECT 1 FROM user_groups ug INNER JOIN contests_groups cg ON ug.group_id = cg.group_id WHERE ug.username=? AND cg.contest_id=?)`,
```

Comment:
The contest can also be `public`. For public contests, everyone should be allowed to register.

So, query can be modified as:
```suggestion
          WHERE EXISTS(SELECT 1 FROM user_groups ug INNER JOIN contests_groups cg ON ug.group_id = cg.group_id WHERE ug.username=? AND cg.contest_id=?) OR EXISTS(SELECT 1 FROM contests WHERE id=? AND public=1)`,
```
And query array would become:
```
[contestId, username, username, contestId, contestId],
```